---
title: "Large datasets assignment"
output: html_document
date: "2024-02-26"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.

```{r}
#Clear the R environment: if anything from a previous session is here it clears the objects that were made before. 
rm(list = ls())

```

```{r}


#Install the package Rcompadre, which will allow you to access COMADRE (and COMPADRE) online
#install.packages("Rcompadre")
#install.packages("tidyverse) # this is for data wrangling

#Call the library so you can start using it:
library(Rcompadre)
library(tidyverse)

#Fetch the most recent database version from www.compadre-db.org with:
comadre <- cdb_fetch("comadre") 
#This function takes comadre from the webiste: there are two data bases it deals with: comadre vs compadre: we are telling it to fetch comadre, which is for mammals (not plants)

#The command 'dim' outputs the dimensions of an object, thus giving you an idea of how many row entries and column variables there are in the original database: tells you the dimensions of the database: 
dim(comadre)


```

Exploring the metadata in comadre. Metadata is the extra bits of information surrounding your data. Eg if your data is heights of different trees, metadata is location and species data 

```{r}
#Explore the metadata in COMADRE
comadre@data
#Choosing the taxonomic subset of animals, reptiles. 

```
```{r}

#Examining all the possible classes in the dataset I could choose from 
unique_classes <- unique(comadre@data$Class)
# Print the unique classes
print(unique_classes)

#Finding the number of unique species in the class Reptilia ?
```


```{r}
#subsetting our data to only include reptiles, and to only include wild individuals, not captive individuals, and species that are not constrained geographically 
comadre1 <- subset(comadre,
                   # Consider only reptiles
                   Class == "Reptilia" & 
                 
                  # studied under "control" (i.e. no treatment) conditions
                   MatrixTreatment == "Unmanipulated" &
                   # and only from "W"ild populations
                   MatrixCaptivity == "W")     

#taking a subset of the whole dataset to make a smaller dataset to make a subset of only the species we are interested in. The criterium being imposed are: just reptiles, with no geographical constraint. Matrix treatement (find out more and go to the URL to find in detail description): dynamics of huge populations under natural conditions. Establishing relationships between generation time and conservation studies we need to ensure that statistical inferences are to do with these and nothing else. We subset the data to allow us to draw fair comparisons? The last criterium is on whether they are from wild populations or not.

#ensuring that there are at least 10 unique species in this subset 

#Examining all the possible classes in the dataset I could choose from 
unique_species_reptilia <- unique(comadre1$Species)
# Print the unique classes
length(unique_species_reptilia)
#There are 28 unique species of reptiles in this dataset. 

#calculating the dimensions of my new subsetted comadre1: 
dim(comadre1)

```

We will use the function cdb_flag() to add columns to the database flagging potential issues with these MPMs, such as missing values, or MPMs that don’t meet assumptions necessary to calculate population growth rates (like ergodicity, irreducibility, or primitivity 

```{r}
#we cannot calculate population growth rate from any data that has NAs attached or is not ergodic. 
#ergodic means that there is statistical equivalence between mean values measured over time and mean values measured across space 

#we are therefore further subsetting the data based of of these requirements 
comadre_flags <- cdb_flag(comadre1)
comadre_sub <- subset(comadre_flags,
                      check_NA_A == FALSE &
                      check_ergodic == TRUE)

#Check size of subset data again:
dim(comadre_sub)
head(comadre_sub) 

```

Three fundamental factors impacting populations are: 
Reproduciton 
Survival 
Growth / development. 
These are things any species needs to do to fulfil its rule to pass on its genes and survive, reproduce and close the loop. 
Reproduction comes in different flavours: sexual vs asexual reproduction (clonality). Survival and growth: parts of the same coin. Growth depends on survival, you cant grow tomorrow if you die today. some species have the ability to decrease in size so shrinkage and growth are part of the same thing. Any population matrix is a three dimensinal layer / cake. 

```{r}
#Explore some MPMs in your subset database
#The first MPM stored in your data is:
matA(comadre_sub)[[1]] #the subset contains many MPMs, this is just one of them 
#Extracts a full matrix population model from a subset of comadre: this is the three layered pancake. We need to know which numbers correspond to which layer. 
#The value that is 76 is reproduction as it cannot be survival because the value is more than 1 whereas survival should be bounded between 0 and 1 

#We are now going to add the submatrices U, F and C: 
matU(comadre_sub)[[1]] #This is the survival component of the matrix, values are between 0 and 1 
matF(comadre_sub)[[1]] #the sexual component of the matrix population model 
matC(comadre_sub)[[1]] #the clonal component of the matrix population model, makes sense that it is all 0s  

comadre_sub@data[1,] #This line shows one portion of the matrix, the first row
```
the population growth rate ($\lambda$) corresponds to the dominant eigenvalue of the MPM *A*, and it projects the rate of growth (if $\lambda$ > 1) or decline ($\lambda$ < 1) of the population. One of the three definitions of generation time (T) is the mean age of reproductive individuals in the population. This life history trait is tightly linked to the rate of population turnover, and it is frequently used in assessments of vulnerability of the IUCN.

1: Show a histogram of population growth rate and a separate histogram of generation time of my subset. 

Generating a histogram for population growth rate 

```{r}
#Obtain population growth rates using the function 
#"lambda" from the R package "popbio
#install.packages("popbio") (need an if loop in the code to install this if it has not yet been installed.)
library("popbio")

#Using lapply (which applies a function to every element in a list), 
#we will calculate the population growth rate (lambda) for each matrix
#looking at the subset: from our subset we call matA and apply the lambda function to it which extracts the growth rate 
comadre_sub$lambda <- unlist(lapply(matA(comadre_sub), popbio::lambda))

#Examine the distribution of population growth rates:
hist(comadre_sub$lambda, xlab= "Population growth rate", main = "Frequency distribution of Population growth rates")
abline(v = 1, col = 'red', lty = 2, lwd = 2) #this histogram shows the frequency distribution of population growth rates
#This line of code adds a verticle line at x = 1. If population growth rate = 1 this means the population is not changing in size and is at demographic equilibrium. If the conditions under which the matrix was parameterized dont change the population will remain stable 
#Populations change over time so if you want to consider a population over time the value of lambda changes, for example this might come from an abnormally good year. if you assume as many resources and that density dependence isnt important (hasnt reached carrying capacity) lambda would be higher than one. So unless you calculate lambda through a lot of matrix following populations over a long time this is problematic. 


#hist(comadre_sub$MatrixID) #this is a historgram showing the distribution of the matrices? 


```
The (mean?) frequency distributoin of the population centers around 1. This makes sense because this means the population is not changing in size and is at demographic equilibrium.


Generating a histogram for distributions of generation time: 

```{r}
#To calculate generation time, we will use the R package "Rage"
#install.packages("Rage")
library("Rage")

#Again using lapply, we will now use the U and F matrices 
#to calculate the generation time
comadre_sub$gen_T <- mapply(Rage::gen_time, matU(comadre_sub), matF(comadre_sub))

#To calculate population growth rate you can claculate this with matrix A. But to calculate generation time you need to separate the processes. So instead of feeding MAT A we need the sub matrix U, survival and growth, and F which includes sexual reproduction. Splitting the matrix up. 

#In the code above, the accessor function matA() is used to extract a list 
#of projection matrices (the full matrix, matA) from every row of comadre_sub. 
#There are also accessor functions for the matrix subcomponents (matU(), 
#matF(), matC()), and for many other parts of the database too.

#Examine the distribution of generation times:
#Generation time is the number of years from birth to reproduction. There are actually three definitions but the most accapted is this. Or can think of as the mean age at which individuals reproduce.

#Why do we care? Important: populations with shorter generation times can respond faster/ evolve faster in response to stress. Species with shorter generation times evolve faster. If you have information about generation times, can you come up with a hypothesis linking conservation status and generation time? 
#We can hypothesize that those with better conservation status have shorter generation times, less endangered.

hist(comadre_sub$gen_T, xlab = "Generation time (years)", main = "Frequency distribution of generation times")
#there is a higher frequency of populations with shorter generation times 

```
There is a higher frequency of populations with shorter generation times. This makes sense because they can evolve faster in response to stress so are less likely to be endangered (we can hypothesize this)


2. Fit a model to explore whether the generation time and population growth rate of your species’ populations are correlated. Would you have expected them to be correlated? Why?

Attempt for question 2: 

```{r}
#The data about population growth rate is: 
comadre_sub$lambda
#The data about generation time is: 
comadre_sub$gen_T



#Checking if there are NAs in lambda values and removing these. Asking out of all the lamdas are there some that are nas? This is a conditional line that gets rid of nas that in this case isn't fulfilled as there isnt nas in this dataset. If you rerun this in another dataset that has nas it would work.
if (length(which(is.na(comadre_sub$lambda)))>0) {
  comadre_sub <- comadre_sub[-which(is.na(comadre_sub$lambda)),]
}

if (length(which(is.na(comadre_sub$gen_T)))>0) {
  comadre_sub <- comadre_sub[-which(is.na(comadre_sub$gen_T)),]
}


if (length(which(is.infinite(comadre_sub$gen_T)))>0) {
  comadre_sub <- comadre_sub[-which(is.infinite(comadre_sub$gen_T)),]
}

```

```{r}
#What kind of model do I want to use? 
#A linear regression analysis 

# Fit a linear regression model
linear_model <- lm(comadre_sub$lambda ~ comadre_sub$gen_T)

# Summarize the model
summary(linear_model)

```
The p value is > 0.05 so there is no statistically significant linear correlation between between generation time and population growth rate.

EXPLAIN WHY YOU WOULD HAVE EXPECTED TO SEE A CORRELATION 


3: 
Obtaining conservation status of these species from the IUCN Red list. Then explore, using statistical methods whether generation time and population growth rate predict conservation status 

Note that the object 'comadre_sub' contains quite a few MPMs for each species. Sometimes these MPMs correspond to different years (e.g., 2000 to 2001, 2001, to 2002...) within the same population, and sometimes they may refer to different populations, or even come from different publications/research groups. As such, for the purposes of today's exercise, and in the interest of time, we will only sample one MPM per species:
```{r, message=FALSE, error=FALSE, warning=FALSE}
#Identifying which species names are unique:
unique_comadre <- unique(comadre_sub$SpeciesAccepted)
#If there are multiple treatments on a species there is more than one matrix
length(unique_comadre)
#Resulting in 25 reptile species 

#We could sample a single MPM wihtin each species in different ways, but here we are going to simply selecting the first entry in the whole of comadre_sub where there is a new species using the command 'duplicated'. This function examines element by element in a vector whether the next element has already appeared before in the vector.
comadre_sub <- comadre_sub[which(duplicated(comadre_sub$SpeciesAccepted)==FALSE),] #after executing this line of code, comadre_sub will only contain unique rows based on the values in the column 'SpeciesAccepted', so each MPM only represents one species 
dim(comadre_sub)



```
Accessing the IUCN data:
```{r}
#Install and call the R package 
#install.packages("taxize")
library(taxize)
#https:/apiv3.iucnredlist.org', provide IUCN access.  ?

API_token <- "8d9c556bb4aa2efeacdc49775c0d0514b42139c96d5ed0f6b9ca5a495ffa49a4"
#Using the API token to extract information from IUCN 
#run the summary from IUCN for our species names
IUCN.list <- iucn_summary(comadre_sub$SpeciesAccepted, distr_detail = FALSE, key = API_token)

```


Carrying out some data wrangling in order to just obtain information on the IUCN redlist status of each of our unique reptile species, of which there are 25. We are then binding this information back onto the comadre data 

Xenosaurus sp was not found in the IUCN database, what does this mean? Because it goes by another name, Zonosaurus sp 

```{r}
#This IUCN list contains a lot of information for each species. We just want 
#their classification and their species name to link to the species
library(tidyverse)

IUCN_data <- lapply(IUCN.list, function(x){
  return(tibble(IUCNstatus = x$status))}) %>% 
  bind_rows() %>% 
  mutate(SpeciesAccepted = comadre_sub$SpeciesAccepted)

#Now link back to original data using the left_join() function
comadre_iucn <- comadre_sub %>%
  left_join(x = ., y = IUCN_data, by = "SpeciesAccepted") %>% 
  mutate(IUCNstatus = case_when(
    IUCNstatus == "EN" ~ "Endangered",
    IUCNstatus == "VU" ~ "Vulnerable",
    IUCNstatus == "NT" ~ "Near Threatened",
    IUCNstatus == "LC" ~ "Least Concern",
    is.na(IUCNstatus) == TRUE ~ "Not Assessed"),
    IUCNstatus = factor(IUCNstatus, 
                        levels = c("Endangered","Vulnerable",
                                   "Near Threatened","Least Concern",
                                   "Not Assessed")))
#whenever it is vulnerable, writing it as Vu, etc. 
dim(comadre_iucn)

#We are going to flatten out this object into a simple data.frame
comadre_iucn <- as.data.frame(comadre_iucn) 

#Adding colours to the IUCN categories
comadre_iucn$iucn_colour <- NA
#Making a variable to accomodate colour 
comadre_iucn$iucn_colour[which(comadre_iucn$IUCNstatus == "Endangered")] <- "Orange" 
comadre_iucn$iucn_colour[which(comadre_iucn$IUCNstatus == "Vulnerable")] <- "Yellow"
comadre_iucn$iucn_colour[which(comadre_iucn$IUCNstatus == "Near Threatened")] <- "Dark green"
comadre_iucn$iucn_colour[which(comadre_iucn$IUCNstatus == "Least Concern")] <- "Green"
comadre_iucn$iucn_colour[which(comadre_iucn$IUCNstatus == "Not Assessed")] <- "Blue"

head(comadre_iucn)


```

Now explore both visually and statistically the relationships between generation time, on the one hand, and population growth rate, on the other hand. 



```{r}
#Exploring the relationship between generation time and population growth rate 
ggplot(comadre_iucn, aes(x = IUCNstatus, y = gen_T, 
                         fill = IUCNstatus)) + 
  geom_boxplot(show.legend = F) +
  geom_jitter(width = 0.2, show.legend = F) +
  scale_fill_viridis_d(option = "B", begin = 0.3, end = 0.8) +
  labs(x = "IUCN redlist status", y = "Generation time (years)") +
  theme_classic(base_size = 14)
#Created a boxplot of generation time against IUCN status. We made a hypothesis on how generation time is related to conservation studies. We predicted that species of shorter generation times should be less likely to be endangered. 

ANOVA_gen_T <- aov(gen_T~IUCNstatus, data = comadre_iucn)
summary(ANOVA_gen_T)
#The p value is > 0.05 so the relationship is not significant 

#Now plotting population growth rate against IUCN status 
ggplot(comadre_iucn, aes(x = IUCNstatus, y = lambda, 
                         fill = IUCNstatus)) + 
  geom_boxplot(show.legend = F) +
  geom_jitter(width = 0.2, show.legend = F) +
  scale_fill_viridis_d(option = "B", begin = 0.3, end = 0.8) +
  labs(x = "IUCN redlist status", y = "Population growth rate") +
  theme_classic(base_size = 14)

ANOVA_lambda <- aov(lambda~IUCNstatus, data = comadre_iucn)
summary(ANOVA_lambda)
#P is also > 0.05 so the relationship is not significant 

```
None of the relationships are statistically significant. Why?? 

Question 4: 
Plotting the values of generation time and population growth rate of your species on the phylogeny provided to you. Is there any visual evidence of phylogenetic inertia in those traits? Why? 

(phylogenetic inertia refers to the limitations on future evolutionary pathways that have been imposed by previous adaptations)

The first step is to run a taxonomic check on your species names, in case these are not taxonomically accepted ("up to date" according the latest biological evidence regarding species relationships)



```{r}



#Once again, state your species names
my_taxa <- comadre_iucn$SpeciesAccepted
#We are checking whether the species are taxonomically accepted 

#Use the command 'tnrs_match_names' to run my_taxa through the API of ROTL to check species names
resolved_names <- rotl::tnrs_match_names(names = my_taxa)
#If the subset you choose is quite big the code should take a bit longer to run but should be fine. 


#Inspect the produced table named "resolved_names", which provides taxonomically accepted names, and additional info, such as the unique ott_id that we can then use to obtain phylogenetic tree information from OTL
resolved_names
#approximate match means: it is true if it finds suggestions of names you might want to consider as an alternative but in this case it is all fine. 
#OTT ID is the unique number given to any species across the tree of life. 
#We need to eliminate any species whose ott_id is NA to proceed with the following steps and then re-do the previous steps:
#the approximate match is also flagged 'incertae sedis' which means its precise location in the tree of life is uncertain / disputed. there is insufficient data to confidently place a taxon within a specific clade. 

#it has therefore been pruned from the latest tree in the open tree of life. I am therefore going to prune it from my data. 

```

```{r}
comadre_iucn$ott_id <- resolved_names$ott_id
#Within this there is a variable called ott_id which we are interested, we are creating a variable called this and saving this information in our data frame 
dim(comadre_iucn)
comadre_iucn <- comadre_iucn[is.na(comadre_iucn$ott_id) == FALSE,]
#Getting rid of any NAs 
dim(comadre_iucn)

```



```{r}
#Removing the species Zonosaurus sp from comadre_iucn because it is incertae sedis so it has been pruned from the current tree of life as its precise position in the open tree of life is uncertain 
# Assuming you have the dplyr package installed, if not, install it using install.packages("dplyr")
library(dplyr)

# Assuming your data frame is named comadre_iucn
# Remove the 25th row
comadre_iucn_pruned <- comadre_iucn %>%
  slice(-25)


```

```{r}

my_taxa <- comadre_iucn_pruned$SpeciesAccepted
#Making a new dataset with the currently accepted name of the species 
resolved_names <- rotl::tnrs_match_names(names = my_taxa)
comadre_iucn_pruned$OTL_unique_name <- resolved_names$unique_name
#Passing on this information to the repository to make sure all these are currently accepted names 

#Adding the names of the species to the data.frame for later merging to the phylogenetic tree with this data.frame
#Data frame you use needs to have species names as row names 
row.names(comadre_iucn_pruned) <- resolved_names$unique_name

#Obtain the phylogenetic tree for your species
#Making a tree from the dataset we have made: the open tree of life package: telling the open tree of life we have unique IDs and take the full phylogeny and return to me the phylogeny containing only the hits corresponding to my mammal species 
my_tree <- rotl::tol_induced_subtree(resolved_names$ott_id)

#the OTL unique name and the species name for OTU 3091 don't match up: the species name is zonosaurus sp but the name we originally searched for was xenosaurus sp.

```

```{r}

#Plot the tree:
plot(my_tree)

#Delete the ott ids for easier plotting:to make it less information dense 
my_tree$tip.label <- gsub('_ott', '', my_tree$tip.label)
#substituting _ott with nothing, ie deleting it. 
my_tree$tip.label <- gsub('[0-9]+', '', my_tree$tip.label)
#substituting any numeric combination between 0 and 9 and all numbers in between with nothing ie deleting all the numbers 
my_tree$tip.label <- gsub('_', ' ', my_tree$tip.label)

my_tree$tip.label

#Plotting my new tree

plot(my_tree, cex = 0.8)

#Plot it in radial format

plot(my_tree, cex = 0.2, type = "fan")

```

Plotting generation time and population growth rate onto a tree 

```{r}
#We will be using the R package phytools
install.packages("phytools")
library(phytools)

#Download the tree, save it in a folder in the computer, and change the writing directory to the appropriate place in the command below before you read the tree in: This phylogeny has a lot of information on plants and animals in comadre and compadre 
setwd("/Users/cissi.user.2/Documents/Computer coding/Large datasets assignment")
tree <- read.tree("COMPADRE-COMADRE_Phylo_June_16_2019.tre")


#This tree contains information for most of the species in COMADRE and COMPADRE, and can also be found here https://github.com/jonesor/compadreDB/tree/master/ under the subfolder "Phylogeny" 
plot(tree)

#Let's clean up the names in the tree, so we can compare them with the names in our object
tree$tip.label <- gsub("_", " ", tree$tip.label)
#Getting rid of underscores for plotting 

#Now we need to "prune" the tree, which means to cut out the branches that are not present in our data:
#The function setdiff: we dont want the whole phylogeny but the part that matches our mammal species. setdiff tells us whih elements in the first vector are not present in the second vector. 
pruned_tree <- drop.tip(tree, setdiff(tree$tip.label, comadre_iucn_pruned$OTL_unique_name))

#The resulting tree has .. tips, meaning .. species. This is just printing what is in our new pruned tree.
pruned_tree
plot(pruned_tree)
#why have I lost a species, there should be 24 tips? 

```
```{r}

# Extract tip labels from the tree object
tree_tip_labels <- tree$tip.label

# Extract tip labels from the comadre_iucn_pruned data frame
comadre_iucn_pruned_tip_labels <- comadre_iucn_pruned$OTL_unique_name

# Compare the two sets of tip labels
matching_labels <- intersect(tree_tip_labels, comadre_iucn_pruned_tip_labels)

# Display the matching tip labels
print(matching_labels)

# Check if there are any tip labels in comadre_iucn_pruned that are not in the tree
missing_labels <- setdiff(comadre_iucn_pruned_tip_labels, tree_tip_labels)
print(missing_labels)
#which is why the tree I plotted only has 23 tips 
#all the labels in comadre_iucn pruned are not in the tree. why? 

# Check if there are any tip labels in the tree that are not in comadre_iucn_pruned
extra_labels <- setdiff(tree_tip_labels, comadre_iucn_pruned_tip_labels)
print(extra_labels)
```
```{r}

#So naturally it seems like we are missing some from our object, which are not present in the tree.?
comadre_iucn_pruned$OTL_unique_name[which(comadre_iucn_pruned$OTL_unique_name%in%pruned_tree$tip.label==F)]
#Locating all the items that are in one of the datasets and not the other one and outputting this. The function %in%: for every slot in the first vector, are you present in the second vector? Asking if all the species are present in the phylogeny. We want to get rid of the species that aren't present, dont want to plot them in the phylogeny. 
#So we now have to elimnate "Canis lupus" and "Theropithecus gelada" from our data to proceed with the next steps:

#output is Scleprous mucronatus mucronatus
#so we have to remove this from the plotting? 
comadre_iucn_pruned <- comadre_iucn_pruned[-which(comadre_iucn_pruned$OTL_unique_name%in%pruned_tree$tip.label==F),]
#We are matching the tip labels of the pruned tree to the unique names of the IUCN list: modifying the subset so the order of species matches the order in the phylogeny so we can map the two against eachother? 
dim(comadre_iucn_pruned)

#Next, we need to reorder the comadre_iucn order so the species appear in the same order in which they appear in the tree
comadre_iucn_pruned <- comadre_iucn_pruned[match(pruned_tree$tip.label,comadre_iucn_pruned$OTL_unique_name),]
dim(comadre_iucn_pruned)

#Checking that the order is the same between the object and the tree
comadre_iucn_pruned$OTL_unique_name
pruned_tree$tip.label
#they are in the same order

# Let us now plot the values of population growth rate at the tips of the phylogeny, and reconstruct the ancestral state of the originating lineange
log_lambdas <- log(setNames(comadre_iucn_pruned$lambda, rownames(comadre_iucn_pruned)))
#setting names and then log transform: creating an object called log lambdas which has the species name and the value of log lambda 

# create "contMap" object, which allows for the reconstruction of likely ancestral traits
#my_tree2<-di2multi(pruned.tree)?

#We are now visually plotting the value of log lambda against the phylogeny of mammal species where values in red correspond to high values of population growth rate and values in y are relatively low. Is there a pattern? Same population growth rate across one clade, or no? Should you expect that performance of a population should be the same for all primates or all wild cats or all corals? Should there be phylogenetic grouping for phylogenetic grouping? 

mammal_contMap_lambda <- contMap(pruned_tree, log_lambdas,plot=FALSE,res=200)
mammal_contMap_lambda <- setMap(mammal_contMap_lambda, c("white","#FFFFB2","#FECC5C","#FD8D3C","#E31A1C"))
plot(mammal_contMap_lambda, fsize=c(0.7,0.8), leg.txt="log(population growth rate)")
par(mar=c(5.1,4.1,4.1,2.1)) ## reset margins to default

#Chelonia mydas and Caretta caretta show high population growth rate and they are in the same clade 

#Let's do the same thing for generation time
log_gen_T <- log(setNames(comadre_iucn_pruned$gen_T, rownames(comadre_iucn_pruned)))
mammal_contMap_gen_T <- contMap(pruned_tree, log_gen_T,plot=FALSE,res=200)
mammal_contMap_gent_T <- setMap(mammal_contMap_gen_T, c("white","light blue","blue","violet","purple"))
plot(mammal_contMap_gent_T, fsize=c(0.7,0.8), leg.txt="log(generation time) (years)")

#You do see grouping here, species that are more closely related have more similar generation times, eg Xenosaurus grandis, Xenosaurus platyceps, Scleroporous arenicolus and Scleroporus grammicus have more similar, shorter generation times, so there is evidence of phylogenetic inertia. Think about a mechanism whereby generation time is phylogenetically conserved! Age of reproduction difference between two groups of individuals it would create reproductive problems. Does this mechanism explain this pattern? yes, speciation would be caused if the age of reproduction of two groups diverged. explaisn why more closely related individulas have similar generatoin times.



```
Question 5 
5. Create a new variable called “Population_performance”. Assign “Decline” or “Increase” to those species whose unique population growth rate is <1 or >1, respectively. This variable is no longer continuous (like population growth rate, lambda was in your original R MarkDown). 

```{r}

# Creating an empty vector to store the values of the new variable
# Creating an empty vector to store the values of the new variable
new_variable <- vector()

# Iterating over each row in the data frame
for (i in 1:nrow(comadre_iucn_pruned)) {
  # Checking the value of lambda for each species
  if (comadre_iucn_pruned$lambda[i] > 1) {
    new_variable <- c(new_variable, 'Increase')
  } else if (comadre_iucn_pruned$lambda[i] < 1) {
    new_variable <- c(new_variable, 'Decline')
  }
}

#need to explain what this for loop is doing 

# Adding the new variable to the data frame
comadre_iucn_pruned$Population_performance <- new_variable

# Displaying the updated data frame
print(comadre_iucn_pruned)
dim(comadre_iucn_pruned)
head(comadre_iucn_pruned)
```
Explore in this handy website (http://blog.phytools.org) a way to plot that new trait on your phylogenetic tree. Then add another column to the phylogeny in such a way that you show the conservation status of those species (collected in previous steps from the IUCN). See picture below for reference/inspiration

Plotting pop_growth_rate onto my phylogenetic tree 



Applying the ACE (Ancestral Character Estimation) function to estimate the ancestral states of the discrete variable: (being above or below the median diversification rate) defined by my split column. This makes diversification rate a binary character: 

```{r}
recon <- ace(comadre_iucn_pruned$Population_performance, pruned_tree, type = "discrete")
#Viewing the likelihood of the ancestral states for all the nodes in the tree
recon
#You can see the likelihood of ancestral states for all nodes by extracting this part of the ace() output
recon$lik.anc

```
Plotting how diversification rates change over the course of the tree

```{r}
plot(pruned_tree, type = "phylogram", show.tip.label = FALSE)
cols <- c(Increase = "blue", Decline = "yellow") # Define colors for categories
tiplabels(pch = 22, cols[comadre_iucn_pruned$Population_performance], cex = 2)
nodelabels(thermo = recon$lik.anc, piecol = cols, cex = 0.75)
axisPhylo()

```
Modifying this code: 

```{r}
library(ape)

# Assuming pruned_tree is your phylogenetic tree object
# Assuming recon is your reconstruction object
# Assuming comadre_iucn_pruned$Population_performance contains categorical values ('Increase' or 'Decline')

# Plotting the phylogram
plot(pruned_tree, type = "phylogram", show.tip.label = FALSE)

# Define colors for categories
cols <- c(Increase = "blue", Decline = "yellow")

# Plot tip labels
tiplabels(pch = 22, col = cols[comadre_iucn_pruned$Population_performance], cex = 2, adj = 0)

# Plot node labels
nodelabels(thermo = recon$lik.anc, piecol = cols, cex = 0.75)

# Add axis
axisPhylo()


```

```{r}
# Assuming pruned_tree is your phylogenetic tree object
# Assuming recon is your reconstruction object
# Assuming comadre_iucn_pruned$Population_performance contains categorical values ('Increase' or 'Decline')

# Plotting the phylogenetic tree with tip labels
plot(pruned_tree, show.tip.label = TRUE)

# Define colors for categories
cols <- c(Increase = "blue", Decline = "yellow")

# Plot tip labels with colors
tiplabels(pch = 22, col = cols[comadre_iucn_pruned$Population_performance], cex = 2, adj = 0)

# Plot node labels
nodelabels(thermo = recon$lik.anc, piecol = cols, cex = 0.75)

# Add axis
axisPhylo()



```
Getting rid of the nodes key?

```{r}

# Assuming you have already loaded the ggphylo package
install.packages("ggphylo")
library(ggphylo)

# Your existing code
plot(pruned_tree, type = "phylogram", label.offset = 0.5)
cols <- c("blue", "yellow")
tiplabels(pch = 22, bg = cols[comadre_iucn$PopulationPerformance], cex = 2)

# Remove internal node labels
tree_without_labels <- tree.remove.node.labels(pruned_tree)

# Plot the modified tree
plot(tree_without_labels, type = "phylogram", label.offset = 0.5)
axisPhylo()
```


```{r}
head(comadre_iucn_pruned)
```

```{r}
library(dplyr)

comadre_iucn_prunedfortree <- comadre_iucn_pruned %>%
  mutate(IUCNstatus = case_when(
    IUCNstatus == "Endangered" ~ "EN",
    IUCNstatus == "Vulnerable" ~ "VU",
    IUCNstatus == "Near Threatened" ~ "NT",
    IUCNstatus == "Least Concern" ~ "LC",
    IUCNstatus == "Not Assessed" ~ "NA",
    TRUE ~ IUCNstatus  # Keep unchanged if not matched
  ))

head(comadre_iucn_prunedfortree)


```
Plotting IUCN status onto my tree 

```{r}
recon2 <- ace(comadre_iucn_pruned$IUCNstatus, pruned_tree, type = "discrete")
#Viewing the likelihood of the ancestral states for all the nodes in the tree
recon2
#You can see the likelihood of ancestral states for all nodes by extracting this part of the ace() output
recon2$lik.anc


```
```{r}
# Assuming pruned_tree is your phylogenetic tree object
# Assuming recon is your reconstruction object
# Assuming comadre_iucn_pruned$Population_performance contains categorical values ('Increase' or 'Decline')
# Assuming comadre_iucn_pruned$IUCNstatus contains categorical values ('EN', 'VU', 'NT', 'LC', 'NA')

# Plotting the phylogenetic tree with tip labels
plot(pruned_tree, show.tip.label = TRUE)

# Define colors for categories
cols <- c(Increase = "blue", Decline = "yellow")
cols_iucn <- c("EN" = "red", "VU" = "orange", "NT" = "green", "LC" = "blue", "NA" = "gray")  # Define colors for IUCN status

# Plot tip labels with colors for Population_performance
tiplabels(pch = 22, col = cols[comadre_iucn_pruned$Population_performance], cex = 2, adj = 0)

# Plot tip labels with colors for IUCNstatus
tiplabels(pch = 22, col = cols_iucn[comadre_iucn_pruned$IUCNstatus], cex = 2, adj = 1)

# Plot node labels for Population_performance
nodelabels(thermo = recon$lik.anc, piecol = cols, cex = 0.75)

# Add additional labels for IUCNstatus
text(x = par("usr")[1] + 0.1, y = par("usr")[3] + 0.1, labels = comadre_iucn_pruned$IUCNstatus, col = cols_iucn[comadre_iucn_pruned$IUCNstatus], cex = 0.5, adj = 0)

# Add axis
axisPhylo()





```